osquery/carver/carver.cpp

```diff
@@ -0,0 +1,304 @@
+                      const std::string& key,
+                      const std::string& value) {
+  std::string carve;
+  auto s = getDatabaseValue(kQueries, kCarverDBPrefix + guid, carve);
+  if (!s.ok()) {
+    VLOG(1) << "Failed to update status of carve in database " << guid;
+    return;
+  }
+
+  pt::ptree tree;
+  try {
+    std::stringstream ss(carve);
+    pt::read_json(ss, tree);
+  } catch (const pt::ptree_error& e) {
+    VLOG(1) << "Failed to parse carve entries: " << e.what();
+    return;
+  }
+
+  tree.put(key, value);
+
+  std::ostringstream os;
+  pt::write_json(os, tree, false);
+  s = setDatabaseValue(kQueries, kCarverDBPrefix + guid, os.str());
+  if (!s.ok()) {
+    VLOG(1) << "Failed to update status of carve in database " << guid;
+    return;
+  }
+}
+
+Carver::Carver(const std::set<std::string>& paths, const std::string& guid) {
+  status_ = Status(0, "Ok");
+  for (const auto& p : paths) {
+    carvePaths_.insert(fs::path(p));
+  }
+
+  // Construct the uri we post our data back to:
+  startUri_ = TLSRequestHelper::makeURI(FLAGS_carver_start_endpoint);
+  contUri_ = TLSRequestHelper::makeURI(FLAGS_carver_continue_endpoint);
+
+  // Generate a unique identifier for this carve
+  carveGuid_ = guid;
+
+  // TODO: Adding in a manifest file of all carved files might be nice.
+  carveDir_ =
+      fs::temp_directory_path() / fs::path(kCarvePathPrefix + carveGuid_);
+  auto ret = fs::create_directory(carveDir_);
+  if (!ret) {
+    status_ = Status(1, "Failed to create carve file store");
+  }
+
+  // Store the path to our archive for later exfiltration
+  archivePath_ = carveDir_ / fs::path(kCarveNamePrefix + carveGuid_ + ".tgz");
+
+  // Update the DB to reflect that the carve is pending.
+  updateCarveValue(carveGuid_, "status", "PENDING");
+};
+
+Carver::~Carver() {
+  fs::remove_all(carveDir_);
+}
+
+void Carver::start() {
+  // If status_ is not Ok, the creation of our tmp FS failed
+  if (!status_.ok()) {
+    return;
+  }
+
+  for (const auto& p : carvePaths_) {
+    if (!fs::exists(p)) {
+      LOG(WARNING) << "File does not exist on disk: " << p;
+    } else {
+      Status s = carve(p);
+      if (!s.ok()) {
+        LOG(WARNING) << "Failed to carve file: " << p;
+      }
+    }
+  }
+
+  std::set<fs::path> carvedFiles;
+  for (const auto& p : platformGlob((carveDir_ / "*").string())) {
+    carvedFiles.insert(fs::path(p));
+  }
+
+  auto s = compress(carvedFiles);
+  if (!s.ok()) {
+    LOG(WARNING) << "Failed to create carve archive: " << s.getMessage();
+    updateCarveValue(carveGuid_, "status", "FAILED");
+    return;
+  }
+
+  s = exfil(archivePath_);
+  if (!s.ok()) {
+    LOG(WARNING) << "Failed to post carve: " << s.getMessage();
+    updateCarveValue(carveGuid_, "status", "FAILED");
+    return;
+  }
+};
+
+Status Carver::carve(const boost::filesystem::path& path) {
+  PlatformFile src(path.string(), PF_OPEN_EXISTING | PF_READ);
+  PlatformFile dst((carveDir_ / path.leaf()).string(),
```

Comment:
Probably not through our generic flow, but I had originally intended the `carve`, `compress`, and `exfill` functions to be made public. As that model's changed we don't exactly need the `!dst.isValid` check below, but I also don't think it's hurting anything having it there.